### ISSUE TITLE
Inconsistent MAPQ values reported by 'view' when MAPQ is masked off

### ISSUE THREAD

**Issue by outpaddling** [samtools/samtools#1329]
When running samtools view on a CRAM file with a required_fields that does not include the MAPQ bit (0x10), most of the reported MAPQ values are 40, but a small percentage are 0.  I would think they would all be reported as a consistent marker value indicating that they are not real data, e.g. '*'.  I was briefly duped into believing I was getting valid MAPQ data by the fact that a believable percentage of them were 0 in the final stats of an analysis.  Due to a typo, the mask was set to 0x208 instead of the 0x218 it should be.

Is this a bug, or an undocumented feature?  If it's the latter, it should probably be clarified in the man page.

samtools 1.10 and 1.11

FreeBSD + clang and CentOS and Amazon Linux + gcc, all x86

samtools view --input-fmt-option required_fields=0x208 NWD104589.b38.irc.v1.cram | awk '{ print $5 }' | more

**Comment by jkbonfield** [comment]
It can depend on how your CRAM file was used.

The required field setting states what fields *must* be decoded, and not what *will* be decoded.  The difference is subtle, but it permits it to avoid decoding things if not needed.  Sometimes however due to the file layout they will still be needed so they'll come back with the correct values anyway.

You can try looking at [io_lib](https://github.com/jkbonfield/io_lib)'s cram_dump and cram_size utilities to get a broad overview of the CRAM file structure.

**Comment by outpaddling** [comment]
Yeah, I understand that fields are grouped in the CRAM containers and one will likely get more info than they requested, but what I failed to make clear initially is that the MAPQ values reported are not correct.  Both the 40s and the 0s appear to be some sort of filler value.

```
FreeBSD barracuda.uits  bacon ~/Data/TOPMed 1007: samtools view --input-fmt-option required_fields=0x218 SRR6990379/NWD104589.b38.irc.v1.cram | mawk '{ print $5 }' | head -200 | tail
28
0
10
0
0
0
0
0
11
20

FreeBSD barracuda.uits  bacon ~/Data/TOPMed 1008: samtools view --input-fmt-option required_fields=0x208 SRR6990379/NWD104589.b38.irc.v1.cram | mawk '{ print $5 }' | head -200 | tail
40
40
40
40
40
40
40
40
40
40
```

**Comment by jkbonfield** [comment]
Oh, that's odd.  I'd expect them to be either filled out with a fixed placeholder value.  Indeed the code states:

```
    if (ds & CRAM_MQ) {
        if (!c->comp_hdr->codecs[DS_MQ]) return -1;
        r |= c->comp_hdr->codecs[DS_MQ]
                        ->decode(s, c->comp_hdr->codecs[DS_MQ], blk,
                                 (char *)&cr->mqual, &out_sz);
    } else {
        cr->mqual = 40;
    }
```

They will get set to 0 if the data is labelled as unmapped.  Hence you may find a mix of 0 and 40, but not randomly so.   Does this tally with your findings?

**Comment by outpaddling** [comment]
The only difference I'm seeing is lack of a CIGAR string where MAPQ is 0.  The reads with MAPQ set to 0 do have a  chromosome and position.

Is that interpreted as unmapped?

Below is the output of 

samtools view -input-fmt-option required_fields=0x208 | awk '$5 == 0' | head -n 4
samtools view -input-fmt-option required_fields=0x208 | awk '$5 == 40' | head -n 4

?	1157	chr1	10135	0	*	*	0	0	GGGTAAGGGGTAGGGTAATGGTTAGGGGTGGGGGGGAGGGGTATGGGATGGGGTTGGGGTGAGGGGAAGGGGGAGGAATGGGGGGAGATGGGATGGGGTGGGGTGGGATTAAGGTGAGGGGGAGGGGAGGGGTTGGGTGGGAGGGATAAGG	*
?	133	chr1	10266	0	*	*	0	0	TGAGACACCAGTAAGAAGGGTTAAGTGTGGAAAACGGGAAAACAAAACCCCTGTGAATCCTGTGCACAGCAGGTGTCCCAAAGCAAGGTGAGAGGCGCAGTGTGAGGTCGACACTGCTGCCTCGAAAGACAAATGTAGTATTCCTAAGGCA	*
?	181	chr1	10287	0	*	*	0	0	TACGAGATAGTGGAGCGAGGGCGGCTGGTAGTGGCGGACGCGACTGGGATCGCCTAGATCTGTACCTGTGGCGGCTGGGTTGGAGTGTGGGGTGGGGGGGGGGGGGAGGGTTTGTGGTGGGGATTGGGCTGGGGTTTGGGTTTGGGGTTGG	*
?	1141	chr1	10377	0	*	*	0	0	GGGGTTTGGGTGTGTTGTTTGTGAGCGGTGGAAGGGGTGGGGGATGGGTGTGGGGTGCGGGTTTGGGTTCGGGTGCTGGGCTGGTCTTGGGGGGGGGGGGGGTGTGGGTATTAGGGTGGGGGGGGGGTTAGGGTTAGGGTTAGGGTTAGGG	*
?	2179	chr1	9999	40	69S60M22S	*	0	0	TTATGAATGCACTAATATCCCCCAAGTATGAGTTGCCTAGAACAGCTCTTCCTTTATTTTCTTTTTCTGGATAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACAGAGATAACTATTGATACAACC	*
?	129	chr1	9999	40	15S60M76S	*	0	0	TATTTTCTTTTTCTGGATAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACAGAGATAACTATTGATACAACACCTTCATGACTCTAACGTACTATCACAGAGTTCTGTGACAAGCTTGCTTTGCCT	*
?	129	chr1	9999	40	46S60M45S	*	0	0	NAGTATGAGTTGCCTAGAACAGCTCTTCCTTTATTTTCTTTTTCTGGATAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACACAGATAACTATTCACACAACAACTTCATGACGCTCCGGTACTAT	*
?	2163	chr1	9999	40	20S60M71S	*	0	0	TCCTTTATTTTCTTTTTCTGGATAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACAGAGATAACTATTGATACAACACCTTCATGACCCTAAGGTACTATCATAGAGTTCTGTGAAACTGATGGTT	*

**Comment by jkbonfield** [comment]
> 
> 
> The only difference I'm seeing is lack of a CIGAR string where MAPQ is 0. The reads with MAPQ set to 0 do have a chromosome and position.

> ```
> ?	1157	chr1	10135	0	*	*	0	0	GGGTAAGGGGTAGGGTAATGGTTAGGGGTGGGGGGGAGGGGTATGGGATGGGGTTGGGGTGAGGGGAAGGGGGAGGAATGGGGGGAGATGGGATGGGGTGGGGTGGGATTAAGGTGAGGGGGAGGGGAGGGGTTGGGTGGGAGGGATAAGG	*
> ?	133	chr1	10266	0	*	*	0	0	TGAGACACCAGTAAGAAGGGTTAAGTGTGGAAAACGGGAAAACAAAACCCCTGTGAATCCTGTGCACAGCAGGTGTCCCAAAGCAAGGTGAGAGGCGCAGTGTGAGGTCGACACTGCTGCCTCGAAAGACAAATGTAGTATTCCTAAGGCA	*
> ?	181	chr1	10287	0	*	*	0	0	TACGAGATAGTGGAGCGAGGGCGGCTGGTAGTGGCGGACGCGACTGGGATCGCCTAGATCTGTACCTGTGGCGGCTGGGTTGGAGTGTGGGGTGGGGGGGGGGGGGAGGGTTTGTGGTGGGGATTGGGCTGGGGTTTGGGTTTGGGGTTGG	*
> ?	1141	chr1	10377	0	*	*	0	0	GGGGTTTGGGTGTGTTGTTTGTGAGCGGTGGAAGGGGTGGGGGATGGGTGTGGGGTGCGGGTTTGGGTTCGGGTGCTGGGCTGGTCTTGGGGGGGGGGGGGGTGTGGGTATTAGGGTGGGGGGGGGGTTAGGGTTAGGGTTAGGGTTAGGG	*
> ```

These are all unmapped.  Use `samtools flags` to decode the flag bits. Eg:

```
@ deskpro107386[nfs_j/jkb]; samtools flags 133
0x85	133	PAIRED,UNMAP,READ2
@ [nfs_j/jkb]; samtools flags 1157
0x485	1157	PAIRED,UNMAP,READ2,DUP
@ [nfs_j/jkb]; samtools flags 181 
0xb5	181	PAIRED,UNMAP,REVERSE,MREVERSE,READ2
@ nfs_j/jkb]; samtools flags 1141
0x475	1141	PAIRED,UNMAP,REVERSE,MREVERSE,READ1,DUP
```

**Comment by outpaddling** [comment]
OK, so the behavior looks consistent.  If you think reporting 0 for unmapped reads and 40 for everything else is reasonable, then I think it just needs to be documented.  I'd suggest a sentence or two in the man page and a code comment explaining why the value 40 was chosen to represent all unknown MAPQs.
